### ISSUE TITLE
Provide a way to use a context when running a process

### ISSUE THREAD

**Issue by aphistic** [go-nacelle/nacelle#5]
In a recent code review with my team one of the questions they brought up was if there was a way to provide a "base context" to each process while it's running. I've thought about it myself at times and I think it would be beneficial to have a context passed in to the `Start()` method on a process. Maybe the `Init()` method too, but I think having an "init context" makes a little less sense than having a "runtime context". I've been trying to think of a way to make the context available in the `Start()` method in a backward-compatible way but so far I haven't actually come up with an idea I like so far and wanted to see what you thought.

The best option, I think, would be to just make `Start() error` into `Start(ctx context.Context) error`, but that obviously wouldn't be backward compatible. Would it really be so bad to make that as a breaking change, though? It's a simple update and maybe Nacelle's ready for v2? :grin: 

The next option I thought of was a `StartWithContext(ctx context.Context) error`, but that runs into the same issue we had when talking about #4, where it modifies the fundamental `Process` interface into an either/or, so I'm pretty sure that's a no-go.

Another idea I had was to do something similar to what we talked about with `Inject(ServiceContainer) error` in #4 would be to create something like a `InjectContext(ctx context.Context)`, or a `BaseContext(ctx context.Context)`. It would feel similar in design to `BaseContext` on `http.Server` to me for setting a base request context, but I have a problem with it not being passed in to the method the context applies to.

So yeah, I'm not sure of a good way to expose that to the process itself. As far as the bootstrap side, I thought there could be both a bootstrap config level and a process level config to set a base context:

```go
ctx := context.Background()

func setupFunc(processes nacelle.ProcessContainer, services nacelle.ServiceContainer) error {
    processes.RegisterProcess(foo.New()) // Uses BaseContext from bootstrapper
    processes.RegisterProcess(bar.New(), nacelle.WithContext(ctx)) // Uses context passed in
}

nacelle.NewBootstrapper("asdf", setupFunc, nacelle.WithBaseContext(ctx))
```

**Comment by efritz** [comment]
I'm super on board with this. I think context popularity didn't happen until after we had the base taxiway/nacelle projects going. No reason not to bump a major version at this point :shrug:.

Might as well make the v2 bump on service as well to take care of the other outstanding issues in an easier way.

**Comment by efritz** [comment]
Do you expect the context to be canceled when the app is shutting down? I think intuitively that should be the case, but I'm not exactly sure when to do so:

- Existing apps use Stop to signal Start to exit, which we could replace entirely with a context canceled by the runner
- Stop can then become an optional interface, as the context replaces 90% of that method's purpose (I believe)
- When Stop is called would you expect Start to still be running? If so we need to cancel the context only after processes have been signaled to stop; if not, then we can cancel it before.

What do you think like the best relation between the context and Start/Stop mechanisms here?

**Comment by aphistic** [comment]
You know, that's a good point... I'm on the fence about canceling the context. On the one hand, it feels like it's not actually what a cancel in a context is for (it's not a client canceling a request) but also who's to decide what a cancel is really for... and maybe it is what a cancel should be for because running the process is being canceled. I'm always so torn on everything related to contexts because I love them so much and find them REALLY useful, but I also want to avoid overusing them and making code harder to maintain overall with them.

I like the idea of making `Stop` an optional interface and it's definitely seemed like more boilerplate than we need to have to implement that part of process management to stop the `Start` method. I had a suggestion originally to allow users to use composition to handle interfaces they didn't want to implement, but as I was playing around with code it didn't work how I wanted.

I ended up experimenting with what I would do if we were to start the nacelle interfaces over and this is an example of what I ended up with: https://repl.it/@aphistic/Nacelle-v2#main.go

It does away with the concept of `Start`/`Stop` and just has a `Run` as the most basic process method, then has some additional interfaces (poorly named on purpose... kinda) to allow a process to handle `Init` before `Run` is called, but also to handle any cases they could do with their own `Stop` method right now with the `Stopping` and `Shutdown` methods. One being called when a process is about to be stopped/canceled and the other for after `Run` returns.

The `processManager` struct itself works very similar to a process right now where `Start` is blocking and `Stop` will tell the process to stop, and `Run` is also blocking. I think it might actually be relatively straightforward to implement?

I think the experiment covers the questions you brought up, let me know if it doesn't or brings up any more! Also, tarmac/taxiway/nacelle definitely started before `context` was even a thing so I think we're good on the reason why it wasn't in the original design. :)
